lib/util/node-event-generator.js

```diff
@@ -37,6 +130,16 @@ class NodeEventGenerator {
      */
     enterNode(node) {
         this.emitter.emit(node.type, node);
+
+        if (node.parent) {
+            this.currentAncestry.unshift(node.parent);
```

Comment:
I'll try it out, but I suspect that this won't help performance since we would need to reverse the list twice every time we add or remove an element.